### PR TITLE
Improve readme (#16) + temp fix for KeyError bug (#17)

### LINKED ISSUE #16: feat: Complete README
I'd suggest adding these to the readme file:
- installation instructions (maybe some badges too, eg [here](https://github.com/bellingcat/vk-url-scraper/))
- development setup (or a link to info on how to setup a poetry project)
- any other cleanup like my addition of the deployment instructions

### PR DESCRIPTION
**Pull request by jclark1913**
This PR expands the readme to include instructions for installing from pip,  using poetry and several other features. It also includes a temporary fix for the KeyError bug mentioned in issue #17 .

Changelog:

#### Readme:

- Add readme template w/ shields
- Add section about Poetry and testing
- Add section for acknowledgements and contributing
- Add install instructions for pip and update commands in readme to prefer using the pypi package over main.py
- Add screenshots
- Add section that explains spreadsheet outputs

#### KeyError bug:
- Rewrite logic in `output.py` using `dict.get()` to avoid KeyErrors. This is a quick fix for now, but the logic may change in the future so I will keep that issue open until I've further reviewed the code.

### CODE REVIEW

**Comment by jclark1913** [comment]
A very small side note: Is there a boilerplate license that Bellingcat likes to use? I selected MIT license, but I notice other projects use different licenses.

File: README.md [hunk 1]
```diff
@@ -35,40 +137,89 @@ The raw output looks something like this:
...
+</div>
+
+Additional libraries/tools: [BeautifulSoup4](https://pypi.org/project/beautifulsoup4/), [Asyncio](https://docs.python.org/3/library/asyncio.html), [Aiohttp](https://docs.aiohttp.org/en/stable/)
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+<!-- GETTING STARTED -->
+## Installation
+
+### Install with pip
```
Comment (msramalho) [inline README.md]:
```suggestion
### Install from [pypi](https://pypi.org/project/wayback-google-analytics/) (with pip)

[![PyPI][pypi-shield]][pypi-url]
```

Hope I got the shield thing right with the links at the end, made 2 suggestions.

File: README.md [hunk 2]
```diff
@@ -101,30 +252,117 @@ options:
...
+<!-- MARKDOWN LINKS & IMAGES -->
+<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
+[contributors-shield]: https://img.shields.io/github/contributors/bellingcat/wayback-google-analytics.svg?style=for-the-badge
+[contributors-url]: https://github.com/bellingcat/wayback-google-analytics/graphs/contributors
+[forks-shield]: https://img.shields.io/github/forks/bellingcat/wayback-google-analytics.svg?style=for-the-badge
+[forks-url]: https://github.com/bellingcat/wayback-google-analytics/network/members
+[stars-shield]: https://img.shields.io/github/stars/bellingcat/wayback-google-analytics.svg?style=for-the-badge
+[stars-url]: https://github.com/bellingcat/wayback-google-analytics/stargazers
+[issues-shield]: https://img.shields.io/github/issues/bellingcat/wayback-google-analytics.svg?style=for-the-badge
+[issues-url]: https://github.com/bellingcat/wayback-google-analytics/issues
```
Comment (msramalho) [inline README.md:364]:
```suggestion
[issues-url]: https://github.com/bellingcat/wayback-google-analytics/issues
[pypi-shield]: https://img.shields.io/pypi/v/wayback-google-analytics
[pypi-url]: https://pypi.org/project/wayback-google-analytics/
```

File: README.md [hunk 3]
```diff
@@ -1,13 +1,115 @@
...
+
+### Why do I need GA codes?
+
+Google Analytics codes are a useful data point when examining relationships between websites. If two seemingly disparate websites share the same UA, GA or GTM code then there is a good chance that they are managed by the same individual or group. This useful breadcrumb has been used by researchers and journalists in OSINT investigations regularly over the last decade, but a recent change in how Google handles its analytics codes threatens to limit its effectiveness. Google began phasing out UA codes as part of its Google Analytics 4 upgrade in July 2023, making it significantly more challenging to use this breadcrumb during investigations.
+
+### How does this tool help me?
+
+Luckily, the Internet Archive's Wayback Machine contains useful snapshots of websites containing their historic GA IDs. While you could feasibly check each snapshot manually, this tool automates that process with the Wayback Machines CDX API to simplify and speed up the process. Enter a list of urls and a time frame (along with extra, optional parameters) to collect current and historic GA, UA and GTM codes and return them in a format you choose (json, txt, xlsx or csv).
```
Comment (msramalho) [inline README.md]:
```suggestion
Luckily, the Internet Archive's [Wayback Machine](https://archive.org/web/) contains useful snapshots of websites containing their historic GA IDs. While you could feasibly check each snapshot manually, this tool automates that process with the Wayback Machines CDX API to simplify and speed up the process. Enter a list of urls and a time frame (along with extra, optional parameters) to collect current and historic GA, UA and GTM codes and return them in a format you choose (json, txt, xlsx or csv).
```
link to the wayback machine if people aren't familiar.

File: README.md [hunk 4]
```diff
@@ -35,40 +137,89 @@ The raw output looks something like this:
     }
 ```
 
-# Installation
+### Further reading
 
-#### Install from source
+- For more info about analytics codes and what the GA-4 rollout means for OSINT: [https://digitalinvestigations.substack.com/p/what-the-rollout-of-google-analytics]
```
Comment (msramalho) [inline README.md]:
```suggestion
- For more info about analytics codes and what the GA-4 rollout means for OSINT: https://digitalinvestigations.substack.com/p/what-the-rollout-of-google-analytics
```
not sure if the brackets were intentional or incomplete markdown, disregard if intentional.

File: README.md [hunk 5]
```diff
@@ -35,40 +137,89 @@ The raw output looks something like this:
...
 
-1. Clone repo:
+- For an example investigation usings analytics codes: [https://www.bellingcat.com/resources/how-tos/2015/07/23/unveiling-hidden-connections-with-google-analytics-ids/]
```
Comment (msramalho) [inline README.md]:
```suggestion
- For an example investigation usings analytics codes: https://www.bellingcat.com/resources/how-tos/2015/07/23/unveiling-hidden-connections-with-google-analytics-ids/
```
not sure if the brackets were intentional or incomplete markdown, disregard if intentional.

File: README.md [hunk 6]
```diff
@@ -35,40 +137,89 @@ The raw output looks something like this:
...
+
+<!-- GETTING STARTED -->
+## Installation
+
+### Install with pip
+
+The easiest way to to install Wayback Google Analytics is from the command line with pip.
+
+1. Open a terminal window and navigate to your chosen directory.
+2. Create a virtual environment and activate it (optional, but recommended)
```
Comment (msramalho) [inline README.md]:
```suggestion
2. Create a virtual environment and activate it (optional, but recommended; if you use [Poetry](https://python-poetry.org/) or [pipenv](https://pipenv.pypa.io/en/latest/) those package managers do it for you)
```

File: README.md [hunk 7]
```diff
@@ -101,30 +252,117 @@ options:
...
+## Development
+
+### Testing
+
+* Run tests with `python -m unittest discover`
+* Check coverage with `coverage run -m unittest`
+
+### Using Poetry for Development
+
+Wayback Google Analytics uses [Poetry](https://python-poetry.org/), a Python dependency management and packaging tool. A github workflow automates the publishing process on pushes to main ((see our workflow here)[https://github.com/bellingcat/wayback-google-analytics/actions]), but be sure to update the version number in `pyproject.toml` when updating.
```
Comment (msramalho) [inline README.md]:
```suggestion
Wayback Google Analytics uses [Poetry](https://python-poetry.org/), a Python dependency management and packaging tool. A GitHub workflow automates the tests on PRs and to main ([see our workflow here](https://github.com/bellingcat/wayback-google-analytics/actions)),  be sure to update the [semantic](https://semver.org/) version number in `pyproject.toml` when opening a PR. 
```

File: README.md [hunk 8]
```diff
@@ -101,30 +252,117 @@ options:
...
+
+To publish a new version to PyPI, you can follow these steps:
```
Comment (msramalho) [inline README.md]:
```suggestion
If you have push access, follow these steps to trigger the GitHub workflow that will build and release a new version to PyPI :
```

File: README.md [hunk 9]
```diff
@@ -101,30 +252,117 @@ options:
...
+
+1. Change the version number in [pyproject.toml](pyproject.toml)
+2. Create a new tag for that version `git tag "vX.0.0"`
+3. Push the tag `git push --tags`
+
+<!-- LICENSE -->
+## License
+
+Distributed under the MIT License. See `LICENSE.txt` for more information.
```
Comment (msramalho) [inline README.md:324]: License file seems to be missing.

File: README.md [hunk 10]
```diff
@@ -101,30 +252,117 @@ options:
...
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+
+
+<!-- CONTACT -->
+## Contact
+
+You can contact me through email or social media.
+
+* email: jclarksummit@gmail.com
```
Comment (msramalho) [inline README.md]:
```suggestion
* email: jclarksummit at gmail dot com
```
not sure how publicly widespread this email address is, but GH is easy to scrape and can easily be used for automated phishing campaigns. do disregard if that's not a concern

File: README.md [hunk 11]
```diff
@@ -101,30 +252,117 @@ options:
...
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+
+
+<!-- ACKNOWLEDGMENTS -->
+## Acknowledgments
+
+* [Bellingcat](https://bellingcat.org) for hosting this project
+* [Miguel Ramalho](https://github.com/msramalho) for constant support, thoughtful code reviews and suggesting the original idea for this project
```
Comment (msramalho) [inline README.md:349]: :rainbow:

**Review by msramalho — CHANGES_REQUESTED** [review]

**Comment by msramalho** [comment]
Had some minor changes and suggestions, otherwise looking good. I had some issue seeing the images on GH but it may be alright when it's merged.

**Pull request merged**